cdap-ui/app/cdap/components/DataPrep/store/DataPrepActionCreator.js

```diff
@@ -76,6 +77,83 @@ export function execute(addDirective, shouldReset, hideLoading = false) {
   });
 }
 
+function setWorkspacePoll(params, observer, workspaceId) {
+  MyDataPrepApi.getWorkspace(params)
+    .subscribe((res) => {
+      let {dataprep} = DataPrepStore.getState();
+      /*
+        1. Open a tab with huge data (like 400 columns and 100 rows)
+        2. Change of mind, open another tab
+        3. 2nd tab's data comes in quick and you are happy browsing it
+        4. Baam 1st tab's data comes and royally overwrites the one you are seeing.
+
+        This is to prevent that. We can't cancel the request we made but we should show only that is relevant
+      */
+      if (dataprep.workspaceId !== workspaceId) {
+        return;
+      }
+      let directives = objectQuery(res, 'values', '0', 'recipe', 'directives') || [];
+      let requestBody = directiveRequestBodyCreator(directives);
+
+      let workspaceUri = objectQuery(res, 'values', '0', 'properties', 'path');
+      let workspaceInfo = objectQuery(res, 'values', '0');
+
+      MyDataPrepApi.execute(params, requestBody)
+        .subscribe((response) => {
+          observer.onNext(response);
+
+          DataPrepStore.dispatch({
+            type: DataPrepActions.setWorkspace,
+            payload: {
+              data: response.values,
+              headers: response.header,
+              types: response.types,
+              directives,
+              workspaceId,
+              workspaceUri,
+              workspaceInfo
+            }
+          });
+
+          fetchColumnsInformation(params, requestBody, response.header);
+
+        }, (err) => {
+          // Backend returned an exception. Show default error message for now able to show data.
+          if (setWorkspaceRetries < 3) {
+            setWorkspaceRetries += 1;
+            setWorkspacePoll(params, observer, workspaceId);
+          } else {
+            observer.onNext(err);
+            DataPrepStore.dispatch({
+              type: DataPrepActions.disableLoading
+            });
+            DataPrepStore.dispatch({
+              type: DataPrepActions.setDataError,
+              payload: {
+                errorMessage: 'Something is wrong'
+              }
+            });
+          }
+        });
+
+    }, (err) => {
+      if (setWorkspaceRetries < 3) {
+        console.log(setWorkspaceRetries);
+        setWorkspaceRetries += 1;
+        setWorkspacePoll(params, observer, workspaceId);
+      } else {
+        console.log('get workspace err', err);
+        DataPrepStore.dispatch({
+          type: DataPrepActions.setDataError,
+          payload: {
+            errorMessage: err.message || err.response.message || err.response
```

Comment:
Would be safer if we didn't `objectQuery` here. Have seen this happen where the `err` from backend is a string and `err.message` is undefined which causes js error because react cannot render `undefined`